pkg/grpc/codec.go

```diff
@@ -0,0 +1,62 @@
+package grpc
+
+import (
+	"google.golang.org/grpc/encoding"
+	"google.golang.org/grpc/encoding/proto"
+	"google.golang.org/grpc/mem"
+)
+
+type vtprotoMessage interface {
+	MarshalToSizedBufferVT(data []byte) (int, error)
+	UnmarshalVT([]byte) error
+	SizeVT() int
+}
+
+type codec struct {
+	// similar to customMarshaler we fall back to original implementation when message is not supported
+	encoding.CodecV2
+}
+
+func (c *codec) Name() string { return c.CodecV2.Name() }
+
+var defaultBufferPool = mem.DefaultBufferPool()
+
+func (c *codec) Marshal(v any) (mem.BufferSlice, error) {
+	m, ok := v.(vtprotoMessage)
+	if !ok {
+		return c.CodecV2.Marshal(v)
+	}
+	size := m.SizeVT()
+	if mem.IsBelowBufferPoolingThreshold(size) {
+		buf := make([]byte, size)
+		_, err := m.MarshalToSizedBufferVT(buf)
+		if err != nil {
+			return nil, err
+		}
+		return mem.BufferSlice{mem.SliceBuffer(buf)}, nil
+	}
+	buf := defaultBufferPool.Get(size)
+	_, err := m.MarshalToSizedBufferVT(*buf)
+	if err != nil {
+		defaultBufferPool.Put(buf)
+		return nil, err
+	}
+	return mem.BufferSlice{mem.NewBuffer(buf, defaultBufferPool)}, nil
+}
+
+func (c *codec) Unmarshal(data mem.BufferSlice, v any) error {
+	m, ok := v.(vtprotoMessage)
+	if !ok {
+		return c.CodecV2.Unmarshal(data, v)
+	}
+	buf := data.MaterializeToBuffer(defaultBufferPool)
+	defer buf.Free()
+	return m.UnmarshalVT(buf.ReadOnlyData())
+}
+
+func init() {
```

Comment:
Is there a way to explicitly register the codec rather than at pkg init ?
Upside : better control on where the codec is used
Downside : has to be called in every location where we want to use the codec

Reply:
We already do it in this package so I think it's fine
https://github.com/stackrox/stackrox/blob/669e11f552076199f02460585ca39f950084389a/pkg/grpc/server.go#L53-L56